Auth building block

jsafoan opened the pull request:
#611   Description
Backwards compatible with previous types of tokens including Rokwire, Phone auth and Shibboleth auth

**Changes in auth-middleware**
The auth-middleware is expecting the following additional environment variables:
AUTH_ISSUER:  This is the issuer that is configured in the auth-building-block.
AUTH_PUBKEYS: The auth service needs a public private RSA Key pair, which is provided as an env variable in the following format. This is parsed as the keyset for decoding in def verify_userauth(..), if the issuer of the token is AUTH_ISSUER as above. 
	
```service:"auth"
keys:[{
  "kty":  ,
  "e":  ,
  "kid":   ,
  "n":  
}]
```

**Changes in Profile service**
Expecting an additional env variable:
AUTH_ISSUER: This is the issuer that is configured in the auth-building-block.

In tokenutils.py, tk_uid and tk_auth are set based on the type of issuer of the token. tk_uid maybe set to uid, Uiuc UIN or phone number according to the issuer. Id Info is then always fetched from the uid field in the token

In profiles.py, we removed the check for token type(uid or phoneauth) in profiles.py, since dataset is always fetched by tk_uid, which is set according to issuer of the token.
uin is set from tk_uid if the token is from shibboleth.

sandeep-ps commented:
@jsafoan, I would like to provide some initial feedback. Could you please resolve the merge conflicts? Also, for documentation purposes, could you please create one or more issues with the details in the PR and then update the CHANGELOG based on that? You could also link them in the PR for future reference. Thanks.

sandeep-ps commented:
@jsafoan, Please let us know when this is ready to be reviewed. Thanks.
CC: @shurwit

sandeep-ps commented:
Hi @bingzhang and @ywkim312, please review this PR and test to make sure that the building blocks that you maintain works fine with the changes to auth middleware library.
Hi @wenjzhu, FYI, I'll test Auth BB and App Config BB.

sandeep-ps commented on lib/auth-middleware/auth_middleware/__init__.py:
```diff
@@ -85,71 +95,93 @@ def authenticate(group_name=None, internal_token_only=False):
     # Otherwise, an error is raised.
-    keys = os.getenv('ROKWIRE_API_KEY').strip().split(',')
+    # keys = os.getenv('ROKWIRE_API_KEY').strip().split(',')
+    keys = os.getenv('ROKWIRE_API_KEY').strip()
```
```suggestion
    keys = os.getenv('ROKWIRE_API_KEY').strip().split(',')
```
ROKWIRE_API_KEY contains a comma separated list of keys. The previous line that was commented out needs to be restored/uncommented.

sandeep-ps commented on lib/auth-middleware/auth_middleware/__init__.py:
```diff
@@ -85,71 +95,93 @@ def authenticate(group_name=None, internal_token_only=False):
+    # keys = os.getenv('ROKWIRE_API_KEY').strip().split(',')
+    api_key = os.getenv('ROKWIRE_API_KEY',
+                        '')
+    # for test_key in keys:
```
Same with the code here too. For backward compatibility, the commented out code needs to be reinstated as currently we are using list of API Keys in ROKWIRE_API_KEY.

sandeep-ps commented on lib/auth-middleware/auth_middleware/__init__.py:
```diff
@@ -32,17 +32,25 @@
+rokwire_events_web_app = 'events_web_app'
+rokwire_app_config_manager_group = 'app_config_manager'
 
 ROKWIRE_EVENT_WRITE_GROUPS = [rokwire_event_manager_group, rokwire_events_uploader, rokwire_events_web_app,
```
`rokwire_event_approvers` and `rokwire_group_admins` variables are not initialzed. These are the corresponding group names: 

```
rokwire_events_approvers = 'urn:mace:uiuc.edu:urbana:authman:app-rokwire-service-policy-rokwire event approvers'	
rokwire_group_admins = 'urn:mace:uiuc.edu:urbana:authman:app-rokwire-service-policy-rokwire groups access'
```

sandeep-ps commented on lib/auth-middleware/auth_middleware/__init__.py:
```diff
@@ -85,61 +95,73 @@ def authenticate(group_name=None, internal_token_only=False):
+            else:
+                is_member_of_key = uiucedu_is_member_of
+
+            if is_member_of_key in id_info:
```
For any specific API endpoint, membership in any one of the required LDAP groups is sufficient. As you can see on the left side, the previous code supports that. For backward compatibility, it needs to be reinstated in some form. For example for POST /events, the user just needs to be member of one of the listed AD groups. https://api-dev.rokwire.illinois.edu/docs/?urls.primaryName=Events%20Building%20Block#/Events/post_events

sandeep-ps commented on profileservice/api/controllers/configs.py:
```diff
@@ -22,17 +22,19 @@
 PROFILE_REST_STORAGE = os.getenv('REST_STORAGE', '/usr/src/app/rest')
 MONGO_PROFILE_URL = os.getenv('MONGO_PROFILE_URL', 'localhost:27017')
-MONGO_PII_URL = os.getenv('MONGO_PII_URL', 'localhost:27017')
+MONGO_PII_URL = os.getenv('MONGO_PROFILE_URL', 'localhost:27017')
```
```suggestion
MONGO_PII_URL = os.getenv('MONGO_PII_URL', 'localhost:27017')
```
In Illinois instance, PII database is separate from Non-PII database marked in `MONGO_PROFILE_URL`. Also, from a Rokwire architecture perspective, we think it's a good practice to separate PII data from Non PII data.

sandeep-ps commented on profileservice/api/controllers/configs.py:
```diff
@@ -22,17 +22,19 @@
-PROFILE_DB_PROFILE_COLL_NAME = 'NonPiiDataset'
-PII_DB_PII_COLL_NAME = 'PiiDataset'
+PROFILE_DB_NAME = os.getenv('PROFILE_DB_NAME', 'profiledb')
+PII_DB_NAME = os.getenv('PROFILE_DB_NAME', 'profiledb')
```
```suggestion
PII_DB_NAME = os.getenv('PROFILE_DB_NAME', 'piidb')
```
This is the currently used PII database name in the Illinois instance.

sandeep-ps commented on profileservice/api/controllers/configs.py:
```diff
@@ -22,17 +22,19 @@
-PII_DB_PII_COLL_NAME = 'PiiDataset'
+PROFILE_DB_NAME = os.getenv('PROFILE_DB_NAME', 'profiledb')
+PII_DB_NAME = os.getenv('PROFILE_DB_NAME', 'profiledb')
+PROFILE_DB_PROFILE_COLL_NAME = 'non_pii'
```
Please reinstate other currently used collection and database names. If needed, we can use environment variables to change these in specific instances.

sandeep-ps commented on lib/auth-middleware/auth_middleware/__init__.py:
```diff
@@ -204,9 +239,27 @@ def verify_userauth(id_token, group_name=None, internal_token_only=False):
+        AUTH_PUBKEYS = os.getenv('AUTH_PUBKEYS', '').strip()
+        AUTH_ISSUER = os.getenv(
+            'AUTH_ISSUER', '').strip()
+        print("AUTH ISSUER: ", AUTH_ISSUER)
```
Please either remove print statement or replace it with logger info or debug statement. Thanks.

sandeep-ps commented:
@ywkim312 and @bingzhang, Please share if you have any other feedbacks. I tested both App Config and the current Auth building block and they are working fine. If things look okay, my goal is to merge this PR and deploy profile, COVID app config, authentication, and logging building blocks to dev instance with the new changes today. Thanks.

sandeep-ps commented:
@shurwit, Could you please take a look at the remaining comments in this PR? Also, please update the CHANGELOG. I will do another round of testing after that. Thanks.

sandeep-ps commented on loggingservice/api/controllers/logs.py:
```diff
@@ -19,9 +19,10 @@
 from flask import request
 
-import controllers.config as cfg
+import controllers.configs as cfg
```
```suggestion
import controllers.config as cfg
```

sandeep-ps commented on loggingservice/api/logging_rest_service.py:
```diff
@@ -14,7 +14,7 @@
 import connexion
 import logging
-import controllers.config as cfg
+import controllers.configs as cfg
```
```suggestion
import controllers.config as cfg
```

sandeep-ps commented on profileservice/api/controllers/configs.py:
```diff
@@ -27,10 +27,12 @@
-PROFILE_DB_NAME = 'profiledb'
-PII_DB_NAME = 'piidb'
+PROFILE_DB_NAME = os.getenv('PROFILE_DB_NAME', 'profiledb')
+PII_DB_NAME = os.getenv('PROFILE_DB_NAME', 'piidb')
```
```suggestion
PII_DB_NAME = os.getenv('PII_DB_NAME', 'piidb')
```
To differentiate between Non PII and PII DB names.

sandeep-ps commented:
@shurwit, I'm re-reviewing/testing this and going through all code changes. I hope to merge this today.

Please update Auth Middleware library README to include all new environment variables being added to the library like AUTH_ISSUER, AUTH_PUBKEYS, etc. If possible, adding a very short description of about these will be helpful. Thanks.

sandeep-ps commented on profileservice/api/controllers/profiles.py:
```diff
@@ -242,13 +252,12 @@ def pii_post():
         auth_resp)
-
-    is_new_entry = False
+    print(tk_uid, tk_name, tk_email, tk_phone, tk_auth)
```
Please remove print statement. Thanks.

sandeep-ps commented on profileservice/api/controllers/profiles.py:
```diff
@@ -304,72 +313,22 @@ def pii_post():
-        try:
-            dataset = mongoutils.get_pii_dataset_from_field('phone', tk_phone)
-
+        dataset = mongoutils.get_pii_dataset_from_field('uid', tk_uid)
```
Since currently `uid` field is not in the database, this will not return profiles that are already in the database. Also, this might eventually result in duplicate PII records. One option would be to set this key name as a config option with the default value being 'uin'. Please also see if there are other better solutions. I would like to consider updating the database as a last resort.

sandeep-ps replied:
@shurwit, tagging in case you missed this comment above. Thanks.

sandeep-ps commented on profileservice/api/controllers/profiles.py:
```diff
@@ -621,19 +588,23 @@ def get_data_list_pid(pid):
-        id_from_db = data_list['phone']
-        if id_from_db == id_string:
-            auth_pass = True
+    id_from_db = data_list['uid']
```
Here also, backwards compatibility needs to be maintained, as `uid` won't be there in existing PII database records.

shurwit replied:
I think these following lines should resolve this issue. Let me know if you think this will be insufficient. Thanks!
```
    id_from_db = data_list['uid']
    if id_from_db == id_string:
        auth_pass = True
    else:
        if 'uin' in data_list:  # Shibboleth ID Token
            # get id info from data_list
            id_from_db = data_list['uin']
            if id_from_db == id_string:
                auth_pass = True
        elif 'phone' in data_list:  # Phone ID Token
            # get phone number from data_list
            id_from_db = data_list['phone']
            if id_from_db == id_string:
                auth_pass = True
```

sandeep-ps replied:
@shurwit, For existing users, `uid` will not be present in `data_list` when GET call is made. As `uid` field is added during POST, this will result in error during GET call. Please see a subset of the stack trace below:

```
  File "rokwire-building-blocks-api/profileservice/api/controllers/profiles.py", line 620, in pii_get
    auth_pass = check_id(auth_resp, data_list[0])
  File "rokwire-building-blocks-api/profileservice/api/controllers/profiles.py", line 594, in check_id
    id_from_db = data_list['uid']
KeyError: 'uid'
```

sandeep-ps replied:
@shurwit, just to clarify. The above exception occurs when GET /profiles/pii is called (for existing users) before POST /profiles/pii. Thanks.

shurwit replied:
Hi @sandeep-ps, I have made the following change which should prevent this error from occurring. Please let me know if you see other issues with this part of the code.
```
- id_from_db = data_list['uid']
+ id_from_db = data_list.get('uid')
```

ywkim312 commented:
There are somethings that I would like to write in here based on my test.

For the access token, I had to swap shibboleth id token to the api-dev end point, which is https://api-dev.rokwire.illinois.edu/authbb/uiuc/swap-token. The header should contain ROKWIRE-API-KEY and Authorization that is a bearer value of shibboleth token.

There are two issuer information, one is for ROKWIRE_ISSUER and the other is AUTH_ISSUER, if these are the same, then the public keys will be ROKWIRE_PUB_KEY so the AUTH_PUBKEYS value shouldn't be matter. 

If these issuers are different then the AUTH_PUBKEYS should be used. The format of the key in the README is like 
keys:[{
  "kty":  ,
  "e":  ,
  "kid":   ,
  "n":  
}]
However, it should be changed like
{
"keys":[{
  "kty":  ,
  "e":  ,
  "kid":   ,
  "n":  
}]
}
Since the code tries to load the key as a json. 

Also, getting this key value was little bit tricky for me. I got it from @sandeep-ps and he took it from the mongodb database. This needs to be explained in more detail

sandeep-ps commented:
@shurwit or @jsafoan, When you get a chance, could you please take a look at the outstanding comments in this PR? We would like to merge this reasonably soon and continue our work on splitting the repositories to be consistent with other backend modules. Thanks.

sandeep-ps commented:
Hi @shurwit, I have completed another round of testing. I thought I would link all the currently outstanding items here to make things easier for you:
1. https://github.com/rokwire/rokwire-building-blocks-api/pull/603#discussion_r583260553
2. https://github.com/rokwire/rokwire-building-blocks-api/pull/603#issuecomment-764987717
3. Your branch is currently out-of-date with the develop branch. Please pull the latest from develop.

Please take a look at these when you get a chance. Thanks.

jsafoan commented:
> Hi @shurwit, I have completed another round of testing. I thought I would link all the currently outstanding items here to make things easier for you:
> 
> 1. [#603 (comment)](https://github.com/rokwire/rokwire-building-blocks-api/pull/603#discussion_r583260553)
> 2. [#603 (comment)](https://github.com/rokwire/rokwire-building-blocks-api/pull/603#issuecomment-764987717)
> 3. Your branch is currently out-of-date with the develop branch. Please pull the latest from develop.
> 
> Please take a look at these when you get a chance. Thanks.

Hi @sandeep-ps, sorry for the delay in responding to this. I added a check for null dataset and updated the format for the key. Also the branch has been updated with changes from develop.

sandeep-ps commented:
Hi @jsafoan, No issues. Is the change for updating the key format pushed to your branch? I could not see the relevant change in this PR. This is the related feedback from Yongwook for reference: https://github.com/rokwire/rokwire-building-blocks-api/pull/603#issuecomment-764987717
If you could also update the README based on this feedback, I think it'll help for future development and testing. Thanks.

sandeep-ps commented:
Closing this PR as the new changes for Core BB support were added through #813. Please reopen PR if needed. Thanks.
